recipes/Voicebank/enhance/SGMSE/train.py

```diff
@@ -0,0 +1,874 @@
+            # Load waveforms
+            sig_noisy = sb.dataio.dataio.read_audio(noisy_wav)
+            sig_clean = sb.dataio.dataio.read_audio(clean_wav)
+
+            orig_len = sig_clean.shape[-1]
+            # Pad if too short
+            if orig_len < target_len:
+                needed = target_len - orig_len
+                left_pad = needed // 2
+                right_pad = needed - left_pad
+                sig_noisy = F.pad(
+                    sig_noisy, (left_pad, right_pad), mode="constant"
+                )
+                sig_clean = F.pad(
+                    sig_clean, (left_pad, right_pad), mode="constant"
+                )
+            # Crop if too long
+            elif orig_len > target_len:
+                if random_crop:
+                    start = np.random.randint(0, orig_len - target_len)
+                else:
+                    start = (orig_len - target_len) // 2
+                sig_noisy = sig_noisy[..., start : start + target_len]
+                sig_clean = sig_clean[..., start : start + target_len]
+
+            # 5) normalize
+            if normalize == "noisy":
+                fac = sig_noisy.abs().max()
+            elif normalize == "clean":
+                fac = sig_clean.abs().max()
+            else:
+                fac = 1.0
+
+            return sig_noisy / fac, sig_clean / fac
+
+        return [wav_pairs]
+
+    # create datasets
+    datasets = {}
+    for split, rc in zip(
+        ["train", "valid", "test"],
+        [random_crop_train, random_crop_valid, random_crop_test],
+    ):
+        pipelines = build_pipeline(rc)
+        json_path = hparams[f"{split}_annotation"]
+        datasets[split] = sb.dataio.dataset.DynamicItemDataset.from_json(
+            json_path=json_path,
+            replacements={"data_root": data_dir},
+            dynamic_items=pipelines,
+            output_keys=["id", "noisy_sig", "clean_sig"],
+        )
+
+    # optional length sorting
+    if hparams["sorting"] in ("ascending", "descending"):
+        datasets["train"] = datasets["train"].filtered_sorted(
+            sort_key="length", reverse=hparams["sorting"] == "descending"
+        )
+        hparams["train_dataloader_opts"]["shuffle"] = False
+
+    return datasets
+
+
+if __name__ == "__main__":
+    cli = argparse.ArgumentParser(add_help=False)
+    cli.add_argument(
+        "--resume",
+        type=str,
+        default="",
+        help="Path to an existing run directory to resume.",
+    )
+    resume_args, remaining = cli.parse_known_args()
+
+    hparams_file, run_opts, overrides = sb.parse_arguments(remaining)
+
+    if resume_args.resume:  # Resume
+        run_dir = Path(resume_args.resume).resolve()
+        hparams_file = run_dir / "hyperparams.yaml"
+        overrides = overrides or ""
+    else:  # New
+        run_name = f"run_{datetime.now():%Y-%m-%d_%H-%M-%S}"
+        overrides = (overrides or "") + f"\nrun_name: '{run_name}'"
```

Comment:
I like the resume mechanism here, nice work! I did get tripped up once while using it (the hparams in the results dir did not have a reference I expected to be there) but I don't think there's anything to fix here necessarily. Maybe a message could be nice that states where the hparams are loaded from.